### PR TITLE
Add Combined NMS

### PR DESCRIPTION
**Pull request by pooyadavoodi**
CombinedNMS op was recently added to TensorFlow.
https://github.com/tensorflow/tensorflow/pull/23567

This PR adds CombinedNMS to the object detection API in TensorFlow.

The main difference between this new op and the old ones is that it does all of NMS in one op.
It's much more efficient, and it's also easier to swap out with a more efficient version of it from TF custom backends such as TensorRT and XLA.

### CODE REVIEW

**Comment by pooyadavoodi** [comment]
+@azaks2 @aaroey

**Comment by aaroey** [comment]
+@pengchongjin

**Comment by dcyoung** [comment]
@pooyadavoodi do you happen to have any ballpark performance metrics comparing inference before and after the change?

**Comment by pooyadavoodi** [comment]
@dcyoung In our inference benchmarks, I can see from 0.8x to 1.2x speedup when we use CombinedNMS. The smaller the batch size, the bigger the speedup. Note that <1x means regression.

Note that CombinedNMS only works on CPU. Someone might implement a GPU version of it in the future to make it much faster. 

TF-TRT will be able to optimize CombinedNMS (even the CPU version) to TensorRT NMS and with that I can see about 2x speedup across the board.

File: research/object_detection/core/post_processing.py [hunk 1]
```diff
@@ -496,3 +496,73 @@ def _single_image_nms_fn(args):
...
+
+  Raises:
+    ValueError: if `q` in boxes.shape is not 1 or not equal to number of
+      classes as inferred from scores.shape.
+  """
+
+  with tf.name_scope(scope, 'CombinedNonMaxSuppression'):
+    nmsed_boxes, nmsed_scores, nmsed_classes, num_detections = tf.image.combined_non_max_suppression(
+            boxes, scores, max_size_per_class, max_total_size, iou_thresh,
+            score_thresh, use_static_shapes)
```
Comment (pkulzc) [inline research/object_detection/core/post_processing.py]: combined_non_max_suppression doesn't really have a input argument like this, right?

File: research/object_detection/core/post_processing.py [hunk 2]
```diff
@@ -496,3 +496,73 @@ def _single_image_nms_fn(args):
...
+                                 change_coordinate_frame=False,
+                                 num_valid_boxes=None,
+                                 masks=None,
+                                 additional_fields=None,
+                                 use_static_shapes=False,
+                                 scope=None):
+  """Multi-class version of non maximum suppression that operates on a batch.
+
+  This op is similar to `multiclass_non_max_suppression` but operates on a batch
+  of boxes and scores. See documentation for `multiclass_non_max_suppression`
```
Comment (pkulzc) [inline research/object_detection/core/post_processing.py]: Are these all simply copy/pasted? You should add some comments about that this method really do here.

File: research/object_detection/core/post_processing.py (hunk 1 above)
Comment (pkulzc) [inline research/object_detection/core/post_processing.py]: combined_non_max_suppression doesn't really have a input argument like this, right?

File: research/object_detection/core/post_processing.py [hunk 3]
```diff
@@ -496,3 +496,73 @@ def _single_image_nms_fn(args):
...
+      classes as inferred from scores.shape.
+  """
+
+  with tf.name_scope(scope, 'CombinedNonMaxSuppression'):
+    nmsed_boxes, nmsed_scores, nmsed_classes, num_detections = tf.image.combined_non_max_suppression(
+            boxes, scores, max_size_per_class, max_total_size, iou_thresh,
+            score_thresh, use_static_shapes)
+
+    batch_nmsed_masks = None
+    batch_nmsed_additional_fields = None
```
Comment (pkulzc) [inline research/object_detection/core/post_processing.py]: Why are these not handled?

File: research/object_detection/core/post_processing.py (hunk 1 above)
Comment (supriyar) [reply research/object_detection/core/post_processing.py]: Which particular input are you referring to? The CombinedNonMaxSuppression has 6 inputs as defined in https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/ops/image_ops.cc#L913

File: research/object_detection/core/post_processing.py (hunk 3 above)
Comment (supriyar) [reply research/object_detection/core/post_processing.py]: Currently the CombinedNonMaxSuppression does not have support to handle masks or additional fields. We can handle this once the kernel is updated to use these fields.

File: research/object_detection/core/post_processing.py (hunk 1 above)
Comment (pkulzc) [reply research/object_detection/core/post_processing.py]: There are 7 input arguments here while CombinedNonMaxSuppression only takes 6.

Comment (supriyar) [reply research/object_detection/core/post_processing.py]: Sorry I missed the last attribute - use_static_shapes is the bool argument pad_per_class.

File: research/object_detection/core/post_processing.py (hunk 2 above)
Comment (pooyadavoodi) [reply research/object_detection/core/post_processing.py]: @pkulzc could you PTAL. I just added a better explanation of what the function does.

File: research/object_detection/core/post_processing.py (hunk 1 above)
Comment (pkulzc) [reply research/object_detection/core/post_processing.py]: Could we use keyword argument for combined_non_max_suppression here?

File: research/object_detection/core/post_processing.py (hunk 3 above)
Comment (pkulzc) [reply research/object_detection/core/post_processing.py]: Could you add a comment mentioning this?

File: research/object_detection/core/post_processing.py [hunk 4]
```diff
@@ -496,3 +496,85 @@ def _single_image_nms_fn(args):
...
+    #TODO supriyar: Set static shapes here?
```
Comment (pkulzc) [inline research/object_detection/core/post_processing.py]: Remove this?

File: research/object_detection/protos/post_processing.proto [hunk 5]
```diff
@@ -20,6 +20,9 @@ message BatchNonMaxSuppression {
 
   // Whether to use the implementation of NMS that guarantees static shapes.
   optional bool use_static_shapes = 6 [default = false];
+
+  // Whether to use the Combined NMS op
+  optional bool combined_nms = 7 [default = false];
```
Comment (pkulzc) [inline research/object_detection/protos/post_processing.proto]: use_combined_nms

File: research/object_detection/protos/faster_rcnn.proto [hunk 6]
```diff
@@ -168,6 +168,9 @@ message FasterRcnn {
   // If True, uses implementation of ops with static shape guarantees when
   // running evaluation (specifically not is_training if False).
   optional bool use_static_shapes_for_eval = 37 [default = false];
+
+  // Whether to use Combined NMS for first stage.
+  optional bool first_stage_combined_nms = 38 [default=false];
```
Comment (pkulzc) [inline research/object_detection/protos/faster_rcnn.proto]: use_combined_nms_in_first_stage

File: research/object_detection/core/post_processing_test.py [hunk 7]
```diff
@@ -1077,6 +1077,53 @@ def test_batch_multiclass_nms_with_additional_fields_and_num_valid_boxes(
...
+    self.assertIsNone(nmsed_additional_fields)
+
+    with self.test_session() as sess:
+      (nmsed_boxes, nmsed_scores, nmsed_classes,
+       num_detections) = sess.run([nmsed_boxes, nmsed_scores, nmsed_classes,
+                                   num_detections])
+      self.assertAllClose(nmsed_boxes, exp_nms_corners)
+      self.assertAllClose(nmsed_scores, exp_nms_scores)
+      self.assertAllClose(nmsed_classes, exp_nms_classes)
+      self.assertAllClose(num_detections, [3, 3])
```
Comment (pkulzc) [inline research/object_detection/core/post_processing_test.py]: assertListEqual

**Comment by pkulzc** [comment]
Pooya, we recently updated the post_processing files and there may have some conflicts with your PR. Could you please resolve them?

**Comment by pooyadavoodi** [comment]
Thanks @pkulzc 
I will check. There is also some problem in the PR that causes accuracy regression on Faster-RCNN and Mask-RCNN (not in SSD). I am looking into that too.

**Comment by pooyadavoodi** [comment]
@pkulzc I have made some restructuring in the code. Could you please review.

FasterRCNN doesn't work with combined_nms in this PR mainly because of unsupported arguments. I have put some checks in the code to avoid calling combined_nms with those arguments, but some of those arguments can't be checked during graph construction because they are TF tensors (for those I print warning messages instead). We can add TF conditionals to the graph to check those if you think that's necessary (assuming they don't hurt the performance).

File: research/object_detection/builders/post_processing_builder.py [hunk 8]
```diff
@@ -88,6 +88,10 @@ def _build_non_max_suppressor(nms_config):
                      'max_total_detections.')
   if nms_config.soft_nms_sigma < 0.0:
     raise ValueError('soft_nms_sigma should be non-negative.')
+  if nms_config.use_combined_nms:
+    if nms_config.use_class_agnostic_nms:
```
Comment (pkulzc) [inline research/object_detection/builders/post_processing_builder.py]: Merge two if clauses?

File: research/object_detection/protos/post_processing.proto [hunk 9]
```diff
@@ -39,6 +39,9 @@ message BatchNonMaxSuppression {
 
   // Soft NMS sigma parameter; Bodla et al, https://arxiv.org/abs/1704.04503)
   optional float soft_nms_sigma = 9 [default = 0.0];
+
+  // Whether to use tf.image.combined_non_max_suppression
```
Comment (pkulzc) [inline research/object_detection/protos/post_processing.proto]: Please add period to the end of your comments (same for your other changes in this PR as well).

File: research/object_detection/builders/post_processing_builder.py (hunk 8 above)
Comment (pooyadavoodi) [reply research/object_detection/builders/post_processing_builder.py]: Done.

File: research/object_detection/protos/post_processing.proto (hunk 9 above)
Comment (pooyadavoodi) [reply research/object_detection/protos/post_processing.proto]: Done.

**Comment by pooyadavoodi** [comment]
@pkulzc I think I've answered all the comments. 
Please let me know what you think.

**Comment by aaroey** [comment]
@pkulzc @pengchongjin could you help to merge this PR? Thanks.

**Pull request merged**